### ISSUE TITLE
Fastcomp vs upstream performance

### ISSUE THREAD

**Issue by merrijo** [emscripten-core/emscripten#18742]
Hello, we've been running under the old fastcomp branch for a while with great performance, and I'm trying to switch over to upstream but running into a performance issue. We've found that upstream is performing ~3.5x *slower* in our use case and I can't explain what's happening.

One difference I've seen is that upstream seems to be repeatedly re-calling the same function whereas fastcomp only makes a single call. I can't figure out how to prevent this from happening.

Here's what we can see in fastcomp, note that we have single, uninterrupted calls:
![image](https://user-images.githubusercontent.com/6571016/218639236-e8a35fc7-d976-4467-a22a-e60a0bd10f0d.png)

However, you can see that upstream seems to be splitting the functions and weaving different calls:
![image](https://user-images.githubusercontent.com/6571016/218639232-8889e1e6-3848-4599-be7e-c8d14ee22891.png)

For our fastcomp builds, we used the following parameters via Makefile:
`--bind -O2 --llvm-opts 2 -s FETCH=1 -s BUILD_AS_WORKER=1 -s ALLOW_MEMORY_GROWTH=1 -s ASSERTIONS=1 -s ABORTING_MALLOC=0 -s TOTAL_MEMORY=250MB -s BINARYEN_TRAP_MODE=clamp -std=c++11`

For upstream, we use the following parameters via CMakeLists:
`-lembind -O3 -flto=thin -std=c++11 -s WASM=1 -s FETCH=1 -s BUILD_AS_WORKER=1 -s ALLOW_MEMORY_GROWTH=1 -s ASSERTIONS=1 -s ABORTING_MALLOC=0 -s TOTAL_MEMORY=250MB`

Is there something I'm missing? I've been digging through the documentation and trying different combinations to try and close the gap, but haven't been successful yet. Any help is appreciated!

**Comment by sbc100** [comment]
Are you also compiling with `-flto=thin`?   Without this flag at compile time that link-time lto flag won't have any effect.

Aside from that I can't see anything obviously that could cause this regression.   

You can build with `--profiling-funcs` to get names in your traces BTW.

**Comment by merrijo** [comment]
Thanks for that! I'll make sure to add to my debugging config.

I make sure to combine build and link flags during linking, but maybe I'm doing it wrong? Here's a sample CMakeLists file that I'm using, for reference:
```
cmake_minimum_required(VERSION 3.10)
project(output)

set(CMAKE_CXX_COMPILER "em++")
set(CMAKE_VERBOSE_MAKEFILE on)

add_executable(output main.cpp)

set(WASM_OPT "-O3")
set(WASM_BUILD "-flto=thin -std=c++11")
set(WASM_LINK "-lembind -s EXPORTED_RUNTIME_METHODS=ccall,cwrap,UTF8ToString,stringToUTF8,lengthBytesUTF8 -s EXPORTED_FUNCTIONS=_malloc,_free,_Destroy,_Initialise,_Detect,_GetMeanGreyValInROI,_InitialiseStructuringElements,_GetValidScore -s WASM=1 -s FETCH=1 -s BUILD_AS_WORKER=1 -s ALLOW_MEMORY_GROWTH=1 -s ASSERTIONS=1 -s ABORTING_MALLOC=0 -s TOTAL_MEMORY=250MB")

message("Building RELEASE")
set(CMAKE_BUILD_TYPE release)
set(CMAKE_CXX_FLAGS_RELEASE "${WASM_OPT} ${WASM_BUILD}")
set_target_properties(output PROPERTIES LINK_FLAGS "${WASM_LINK}")
```

I don't think I've missed any required parameters... I also tried passing the link parameters during build, but it throws a lot of warnings about ignored flags so I just left it out.

**Comment by sbc100** [comment]
Yup, that looks like you are doing everything right.  If you could share a full example perhaps somebody could dig deeper into where the differences are here.   What kind of workload is your program performing?

Does adding `--profiling-funcs` to link flags help you identify where the delta is?

**Comment by merrijo** [comment]
It's doing image processing to detect objects and perform some pre-processing. Lots of `std::vector` calls and shuffling data around temporary buffers, both pre-allocated and dynamic.

Adding `--profiling-funcs` added the names, and it's weird -- when upstream breaks up the main `detect` function call, there's lower level functions being weaved through. The following are the unique functions I can see, but are called several times:
- __addtf3
- __floatunsitf
- <small gap in calls, it just ... stopped for a brief few ms?>
- __floatsitf
- __divtf3
- __multf3

ending with `wasm-to-js:i:-_emval_incref`, which I assume is normal. I don't see the other function calls happening on fastcomp, so it must be some lower level optimisation that I can't control? I tried playing around with some clang `#pragma` settings, but they didn't do anything.

I'm not sure if I can provide a full working example, but our `detect` method is something like this:
```c++
int Detect(unsigned char *in_array) {
  img_buff->SetData(in_array);
  detect_data->DetectFromRoi(img_buff);
  validScore = -1.0;
  if (detect_data->GetTopRegion().size() > 10) {
    validScore = detect_data->GetValidScore();
  }
  return detect_data->GetTop().size() + detect_data->GetMiddle().size() + detect_data->GetBottom().size();
}
```

where `img_buff` and `detect_data` are initialized when the WASM loads.

**Comment by sbc100** [comment]
@dschuff @kripken any ideas?

**Comment by sbc100** [comment]
Oh those `tf3` functions I think are for emulation of quad precision floats, which would certainly explain the issue.   Are you using `long double` in your code? 

Under the llvm backend `long double` ends up be fp128 which get emulated on software.   Is there anyway you can try the codebase using `double` rather than `long double`?  at least so that we can confirm this is indeed the issue here.

**Comment by sbc100** [comment]
See https://github.com/emscripten-core/emscripten/blob/main/system/lib/compiler-rt/lib/builtins/addtf3.c#L9

**Comment by merrijo** [comment]
Aha! That's probably it. Some of our calculations are using `long double` for the added precision. I'll try to do a mod with just `double` to see if that's the underlying cause.

**Comment by merrijo** [comment]
Yep, confirmed the issue was repeated use of `long double` during our pre-processing. Thanks for the help!

Are there any plans to properly support `long double` in the future? Or will it remain under software emulation?

**Comment by sbc100** [comment]
There are no immediate plans.  It would most likely require added new f128 instruction to WebAssembly, and some hardware that supports such things.  I imagine it will be while before that happens.

**Comment by dschuff** [comment]
Going back to the "fastcomp vs upstream" issue, let's not forget that the reason fastcomp is faster here is because in fastcomp long doubles are literally just defined as doubles. There is no extra precision.
So if that behavior was good enough for you, then switching to doubles (at least on wasm) should work too.

Having said that, it might be possible to make it easier to do that. I just found out that clang has a flag that does this on [x86](https://clang.llvm.org/docs/ClangCommandLineReference.html#long-double-flags) so maybe that or something like it could be made to work for wasm too.

**Comment by merrijo** [comment]
Oddly enough, when running in fastcomp we were having accuracy issues that were solved by switching `long double`. When I revert to `double` using upstream we started having accuracy issues again...
